pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,181 @@
+	if err := v.cmk.ValidateZonePresent(ctx, datacenterConfig.Spec.Zone); err != nil {
+		return err
+	}
+	logger.MarkPass("Datacenter validated")
+
+	return nil
+}
+
+// TODO: dry out machine configs validations
+func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, cloudStackClusterSpec *spec) error {
+	var etcdMachineConfig *anywherev1.CloudStackMachineConfig
+
+	if len(cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host) <= 0 {
+		return fmt.Errorf("cluster controlPlaneConfiguration.Endpoint.Host is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Domain) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig domain is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Zone.Value) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig zone is not set or is empty")
+	}
+	if len(cloudStackClusterSpec.datacenterConfig.Spec.Network.Value) <= 0 {
+		return fmt.Errorf("CloudStackDatacenterConfig network is not set or is empty")
+	}
+	if cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for control plane")
+	}
+
+	if cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return fmt.Errorf("must specify machineGroupRef for worker nodes")
+	}
+
+	controlPlaneMachineConfig := cloudStackClusterSpec.controlPlaneMachineConfig()
+	if controlPlaneMachineConfig == nil {
+		return fmt.Errorf("cannot find CloudStackMachineConfig %v for control plane", cloudStackClusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name)
+	}
+
+	if cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef == nil {
+			return fmt.Errorf("must specify machineGroupRef for etcd machines")
+		}
+		etcdMachineConfig = cloudStackClusterSpec.etcdMachineConfig()
+		if etcdMachineConfig == nil {
+			return fmt.Errorf("cannot find CloudStackMachineConfig %v for etcd machines", cloudStackClusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name)
+		}
+		if etcdMachineConfig.Spec.Template != controlPlaneMachineConfig.Spec.Template {
+			return fmt.Errorf("control plane and etcd machines must have the same template specified")
+		}
+	}
+
+	if cloudStackClusterSpec.datacenterConfig.Namespace != cloudStackClusterSpec.Namespace {
+		return fmt.Errorf(
+			"CloudStackDatacenterConfig and Cluster objects must have the same namespace: CloudstackDatacenterConfig namespace=%s; Cluster namespace=%s",
+			cloudStackClusterSpec.datacenterConfig.Namespace,
+			cloudStackClusterSpec.Namespace,
+		)
+	}
+
+	workerNodeGroupMachineConfig, ok := v.machineConfigs[cloudStackClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name]
```

Comment:
I think we should avoid this kind of thing
Even if we don't support multiple worker node groups (if we don't, we should probably add a validation for that), there is no reason why we can't run the validations for all the worker node group machine configs in a loop. It takes the same amount of time and it facilitates adding support later